rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,834 @@
+        if is_latest || manifest.timestamp() >= self.before {
+            self.process_valid_manifest(&manifest, path).await?;
+            return Ok(());
+        } else {
+            let mut old_manifests = self.old_manifests.lock().unwrap();
+            old_manifests.push(path.clone());
+        }
+        Ok(())
+    }
+
+    async fn process_valid_manifest(
+        &self,
+        manifest: &Manifest,
+        manifest_path: &Path,
+    ) -> Result<()> {
+        // Read the indexes first as it is async and we want to avoid holding the referenced_files
+        // lock across an await point.
+        let indexes =
+            read_manifest_indexes(&self.dataset.object_store, manifest_path, manifest).await?;
+
+        let mut referenced_files = self.referenced_files.lock().unwrap();
+        for fragment in manifest.fragments.iter() {
+            for file in fragment.files.iter() {
+                let full_data_path = self.dataset.data_dir().child(file.path.as_str());
+                let relative_data_path = remove_prefix(&full_data_path, &self.dataset.base);
+                referenced_files.data_paths.insert(relative_data_path);
+            }
+            let delpath = fragment
+                .deletion_file
+                .as_ref()
+                .map(|delfile| deletion_file_path(&self.dataset.base, fragment.id, delfile));
+            if delpath.is_some() {
+                let relative_path = remove_prefix(&delpath.unwrap(), &self.dataset.base);
+                referenced_files.delete_paths.insert(relative_path);
+            }
+        }
+        if let Some(relative_tx_path) = &manifest.transaction_file {
+            referenced_files
+                .tx_paths
+                .insert(Path::parse("_transactions")?.child(relative_tx_path.as_str()));
+        }
+
+        for index in indexes {
+            let uuid_str = index.uuid.to_string();
+            referenced_files.index_uuids.insert(uuid_str);
+        }
+        Ok(())
+    }
+
+    async fn delete_unreferenced_files(&self) -> Result<RemovalStats> {
+        let removal_stats = Mutex::new(RemovalStats::default());
+        let unreferenced_paths = self
+            .dataset
+            .object_store
+            .read_dir_all(&self.dataset.base, Some(self.before))
+            .await?
+            .try_filter_map(|path| future::ready(self.path_if_not_referenced(path, &removal_stats)))
+            .boxed();
+
+        let old_manifests = self.old_manifests.lock().unwrap().clone();
+        let num_old_manifests = old_manifests.len();
+        let old_manifests_stream = stream::iter(old_manifests).map(Result::<Path>::Ok).boxed();
+
+        let all_paths_to_remove =
+            stream::iter(vec![unreferenced_paths, old_manifests_stream]).flatten();
+
+        self.dataset
+            .object_store
+            .remove_stream(all_paths_to_remove.boxed())
+            .for_each(|_| future::ready(()))
```

Comment:
Ok.  I updated it to go ahead and fail on the first error.  The stream returned from object_store is a concurrent stream (buffered) so it may not bail on exactly the first error.  This could potentially lead to a situation where old versions are "corrupt" (e.g. if the old data files were deleted but for some reason the manifest was not correctly deleted) but without some kind of "atomic batch delete" I don't know that this can be avoided.

Reply:
I wonder if it's worth testing that we can still run this even if some of the files are missing. I think as long as we can still call cleanup on a corrupted version to finish the job, this should be fine.

Reply:
I've added a test for this (at least the "manifest file missing its dataset files" case)